ambry-store/src/main/java/com/github/ambry/store/StatsBasedCompactionPolicy.java

```diff
@@ -121,23 +191,42 @@ private CostBenefitInfo getBestCandidateToCompact(NavigableMap<LogSegmentName, L
    */
   private CostBenefitInfo getCostBenefitInfo(LogSegmentName firstLogSegmentName, LogSegmentName lastLogSegmentName,
       NavigableMap<LogSegmentName, Long> validDataSizePerLogSegment, long segmentCapacity, long segmentHeaderSize,
-      long maxBlobSize) {
+      long maxBlobSize, boolean weightOnBenefit) {
     // cost = total valid data size (or data that needs to be copied over)
     // benefit = no of segments reclaimed = (total source segments - total target segments)
     // total target segments = Ceil (total cost / (segment capacity - segment header - max blob size))
     long totalCost = 0;
+    long adjustedTotalCost = 0;
     int totalSegmentsToBeCompacted = 0;
     LogSegmentName curSegmentName = firstLogSegmentName;
+    // weight more on benefits, don't let one single log segment with extreme small valid data leads to small compactions.
+    // Be default, the storeStatsBasedCompactionMinCostInPercentage is 0.06. 16G*0.06=1G
+    long minSegmentCost = storeConfig.storeStatsBasedCompactionMinCostInPercentage == 0 ? 0
+        : (long) (segmentCapacity * storeConfig.storeStatsBasedCompactionMinCostInPercentage);
     while (!curSegmentName.equals(lastLogSegmentName)) {
       totalCost += validDataSizePerLogSegment.get(curSegmentName);
+      adjustedTotalCost += validDataSizePerLogSegment.get(curSegmentName) < minSegmentCost ? minSegmentCost
```

Comment:
I don't follow this logic. So here we are saying that if valid data to copy is less than 1GiB, then take the cost as 1GiB. Doesn't this penalize segments that are smaller?

Reply:
Yes, the logic is just as you mentioned.
If we set the threshold to 1GB, it means it worths to compact the log segments have less than 1GB valid size. It won't penalize the smallest log segments. It gives chance for the smaller but not smallest log segments to get compacted together with the smallest log segments.

In our current design, if we have log segments as below, (below is the valid size percentage). With current IO effort formula, we'll only compact log segment 3rd and 4th since their valid size is extreme small which is 1% and 1%.
```
double[] validPercentage = {0.9, 0.8, 0.05, 0.01, 0.01, 0.05, 0.1, 0.2, 0.4, 0.7}; 
                                                 1st, 2nd, 3rd, 4th,  ...............  
```

**But actually when the valid size percentage is less than some threshold, they are all ready to get compacted.** 

In the above example, if we set the threshold to 6%, with the adjustedTotalCost tuning, we'll compact log segment 2,3,4 and 5. The compaction benefit is 3, much better than the existing one which is 1. The IO cost added is very small.